core_data_modules/traced_data/io.py

```diff
@@ -439,6 +445,305 @@ def import_coda_to_traced_data_iterable_as_matrix(cls, user, data, key_of_raw, c
+        :type message_id_key: str
+        """
+        for td in data:
+            if raw_key in td:
+                td.append_data(
+                    {message_id_key: SHAUtils.sha_string(td[raw_key])},
+                    Metadata(user, Metadata.get_call_location(), time.time())
+                )
+
+    @staticmethod
+    def _assert_uniquely_coded(data, message_id_key, coded_keys):
+        """
+        Checks that all objects with the same message id have been assigned the same codes for each coded_key in
+        coded_keys.
+
+        Fails with an AssertionError if there are TracedData objects with the same message id but different codes
+        assigned, otherwise has no side-effects.
+
+        :param data: Data to check.
+        :type data: iterable of TracedData
+        :param message_id_key: Key in TracedData objects of the message id.
+        :type message_id_key: str
+        :param coded_keys: Keys in the TracedData objects containing coded data.
+        :type coded_keys: iterable of str
+        """
+        seen_message_ids = dict()  # of message_id -> coded_key -> set of code_ids
+        for td in data:
+            if td[message_id_key] in seen_message_ids:
+                # This message id has been seen before, so check that the codes are the same.
+                seen_code_ids = seen_message_ids[td[message_id_key]]
+                for coded_key in coded_keys:
+                    err_string = "Messages with the same id ({}) have different " \
+                                 "labels for coded_key '{}'".format(td[message_id_key], coded_key)
+
+                    if coded_key not in td:
+                        assert seen_code_ids[coded_key] is None, err_string
+                    elif type(td[coded_key]) == dict:
+                        assert seen_code_ids[coded_key] == td[coded_key]["CodeID"], err_string
+                    else:
+                        assert seen_code_ids[coded_key] == {label["CodeID"] for label in td[coded_key]}, err_string
+            else:
+                # This is a new message id, so add these codes to seen_message_ids
+                new_code_ids = dict()
+                for coded_key in coded_keys:
+                    if coded_key not in td:
+                        new_code_ids[coded_key] = None
+                    elif type(td[coded_key]) == dict:
+                        new_code_ids[coded_key] = td[coded_key]["CodeID"]
+                    else:
+                        new_code_ids[coded_key] = {label["CodeID"] for label in td[coded_key]}
+                seen_message_ids[td[message_id_key]] = new_code_ids
+
+    @staticmethod
+    def _deduplicate_data(data, message_id_key, creation_date_time_key):
+        """
+        De-duplicates data.
+
+        Items in data are considered duplicates if they have the same message id.
+        Where duplicates are found, only the object with the oldest creation date is included in the returned list.s
```

Comment:
This feels like it makes more 'filter_duplicates' than 'deduplicate' in that it isn't mutating the object?

Reply:
changed name + docstring